kubeshop/monokle-core
# Try changesets and release validation v0.8

This PR introduces changesets and releases validation v0.8. The changesets works quite smooth imo so give it a try!

What is changesets?
- [Changesets](https://github.com/changesets/changesets#readme): A way to manage your versioning and changelogs with a focus on monorepos. It's the recommended tool of Turbo (our monorepo build tool): https://turbo.build/repo/docs/handbook/publishing-packages/versioning-and-publishing.

Alternatives considered
- [np](https://github.com/sindresorhus/np) which has no focus on monorepos.

How does it work?

_All of these commands are used on the repository's root. You should NEVER need to go to one of the sub-packages._

TLDR

```
# This you do before each PR
npx changeset # Ask desired bump (major|minor|patch) and changelog message
git commit && git push

# This you can automate in a pipeline
npx changeset version # updates changelog.md, package.json and creates git tag
npm install && npm run build
npx changeset publish --otp=YOUR_OTP_TOKEN
git push --follow-tags
```

In steps:

1. **Add changelog entry with `npx changeset`** Every time you want to add an entry for the **next, future** changelog. Generally this is before each PR. It will prompt you through questions. The end result is a temporary file that 1) stores the desired version bump, and 2) the message.

Example:
```
# .changeset/tame-taxi-heal.md (a temporary file that you commit)
---
"@monokle/validation": minor
---

Add methods to transfer state
```

2. **Bump version and create changelog with `npx changeset version`.** You use this every time you want to create a release. It takes all temporary files and creates a changelog for you and uses the metadata to decide the correct next version.

3. Review and commit changes

4. Ensure up to date dependencies and create a build with `npm install && npm run build`.

5. **Publish to npm with `npx changeset publish --otp={one-time-password}**. This publishes all packages that do not have a release yet.
